### ISSUE TITLE
Referencing Issue "CSV Data Import in GenomicsDB"

### ISSUE THREAD

**Issue by nalinigans** [GenomicsDB/GenomicsDB#25]
Referencing [Issue 227](https://github.com/Intel-HLS/GenomicsDB/issues/227) as development has moved [here](https://github.com/GenomicsDB/GenomicsDB.git).

**Comment by nalinigans** [comment]
@mfawzysami, have you read through the [wiki]( https://github.com/GenomicsDB/GenomicsDB/wiki/Importing-CSVs-into-GenomicsDB)?

**Comment by mfawzysami** [comment]
@nalinigans I have read and created the corresponding json files as stated in the wiki but still i am facing the following problem:

====== config.json========================
{
    "row_based_partitioning" : false,
    "produce_combined_vcf": true,
    "produce_tiledb_array" : true,
    "column_partitions" : [
        {"begin": 0, "workspace":"/rnd/home/mfawzy/projects/gdata/trial/ws", "array": "test0" },
        {"begin": 10000, "workspace":"/rnd/home/mfawzy/projects/gdata/trial/ws", "array": "test1" }
    ],
    "vid_mapping_file" : "/rnd/home/mfawzy/projects/gdata/trial/vidmapping.json",
    "callset_mapping_file" : "/rnd/home/mfawzy/projects/gdata/trial/callsets.json",
    "size_per_column_partition": 10000,
    "treat_deletions_as_intervals" : true,
    "num_parallel_vcf_files" : 1,
    "delete_and_create_tiledb_array" : true,
    "fail_if_updating": false,
    "reference_genome" : "/rnd/home/mfawzy/projects/gdata/trial/chrY.fa",
    "do_ping_pong_buffering" : true,
    "offload_vcf_output_processing" : true,
    "discard_vcf_index": true,
    "compress_tiledb_array" : true,
    "disable_synced_writes" : true,
    "segment_size" : 1048576,
    "num_cells_per_tile" : 1000,
    "ignore_cells_not_in_partition": false

}

========== vidmapping.json =======================
{
    "contigs": {
        "24": {
            "length": 57227415,
            "tiledb_column_offset": 0
        }
    },
    "callset_mapping_file" : "/rnd/home/mfawzy/projects/gdata/trial/callsets.json",
    "fields" : {
        "END":{ "vcf_field_class":["INFO"], "type":"int" },
        "BaseQRankSum":{ "vcf_field_class" : ["INFO"], "type":"float" },
        "DP": { "vcf_field_class":["INFO","FORMAT"], "type":"int" },
        "SB":{ "vcf_field_class" : ["FORMAT"], "type":"int", "length":4 },
        "GT": { "vcf_field_class":["FORMAT"], "type":"int", "length":"PP" },
        "AD": { "vcf_field_class":["FORMAT"], "type":"int", "length":"R" },
        "PL": { "vcf_field_class":["FORMAT"], "type":"int", "length":"G" }
    }
}

========================== callsets.json =============================

{
    "callsets" : {
        "HG00141" : {
            "row_idx" : 0,
            "idx_in_file": 0,
            "filename": "/rnd/home/mfawzy/projects/gdata/trial/ALL.chrY.phase3_integrated_v2a.20130502.genotypes.vcf.gz"
        }
    }
}

===============================================================
I am only using chrY vcf compressed in gz and i have the index .bai file for this particular chrY vcf, when i run the following command , i get the following output 

> vcf2tiledb config.json

##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">
##INFO=<ID=END,Number=1,Type=Integer,Description="Stop position of the interval">
##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description="BaseQRankSum">
##INFO=<ID=DP,Number=1,Type=Integer,Description="DP">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="DP">
##FORMAT=<ID=SB,Number=4,Type=Integer,Description="SB">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype",Description="GT">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="AD">
##FORMAT=<ID=PL,Number=G,Type=Integer,Description="PL">
##contig=<ID=24,length=57227415>
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	HG00141
========================================================================

and i have a folder created named 'ws' but when i issue du -h for this particular folder , i receive the following output:

0	ws/test0/genomicsdb_meta_dir
0	ws/test0
0	ws

==========================================================================

what possibly be the problem

**Comment by mlathara** [comment]
@mfawzysami so this part of your loader configuration file creates two (column) partitions:

```
{"begin": 0, "workspace":"/rnd/home/mfawzy/projects/gdata/trial/ws", "array": "test0" },
{"begin": 10000, "workspace":"/rnd/home/mfawzy/projects/gdata/trial/ws", "array": "test1" }

```
When you run vcf2tiledb without any rank parameter it will load the first/rank0 partition only. See this [link](https://github.com/GenomicsDB/GenomicsDB/wiki/Importing-VCF-data-into-GenomicsDB#running-the-program) for more details.

Is it possible that your VCF doesn't have any variants in the first partition (i.e., the first 10000 positions)? you can either explicitly load the second partition as well by running

```
vcf2tiledb -r 1 config.json
```

Or by changing your config.json to only have a single partition and re-running your existing command

```
{"begin": 0, "workspace":"/rnd/home/mfawzy/projects/gdata/trial/ws", "array": "test0" }
```

**Comment by mfawzysami** [comment]
@mlathara Thank you for your answer, I did what you have said and the data was insert successfully, now i would like to query results in GenomicsDB, i did the query.json file using the same config.json above but nothing is returned, i tried multiple intervals for columns but nothing was returned, just empty json variant calls.
{
    "workspace" : "....../ws/",
    "array" : "test0",
    "query_column_ranges" : [[[ 0,100 ],1]],
    "reference_genome":"..../hg19.fasta",
    "vid_mapping_file": "...../vidmapping.json",
    "callset_mapping_file": "..../callsets.json",
    "query_attributes" : [ "REF", "ALT", "CHROM" ]
}

I have placed dots in place of the full absolute file paths.

Each time i run gt_mpi_gather command, i receive the following results:

{
    "variant_calls": [

    ]
}


What could possibly be the problem !

**Comment by mlathara** [comment]
@mfawzysami so the column ranges you are querying here don't have any variants. You are only querying positions [0,100] (inclusive) and then position 100 again. Based on your vid mapping from earlier, this only maps to contig 24 (chr Y?) positions 1 to 101 (note that positions in the config jsons are 0-based [i.e., start from 0] while VCFs use 1-based [start from 1]).

Based on your previous question, it seems the data you are ingesting doesn't have any variants in the first 10000 positions (at least), so I'd try changing the `query_column_ranges` to a range that does include variants. For example, you could just make it as large as the contig to dump all the variants you've ingested:

```
"query_column_ranges" : [[[ 0,57227415]]]
```
You can add the `-l config.json --produce-Broad-GVCF` flags to generate a GVCF. See more details [here](https://github.com/GenomicsDB/GenomicsDB/wiki/Querying-GenomicsDB)

**Comment by kgururaj** [comment]
We should probably edit the wiki to show a better example (preferably with contig names and ranges instead of column ranges). This issue has been seen many times and my poor example is the primary cause.

**Comment by mfawzysami** [comment]
@kgururaj Yes Please, The Wiki is very bad frankly, I am lost in it and it has some English grammar mistakes, if I go to TileDB their documentation is Excellent but GenomicsDB is a seriously bad wiki.

**Comment by kgururaj** [comment]
Could you point us to other places in the wiki where you had trouble understanding? This will help us improve it. The example you pointed us to was a good case in point which we should fix. Thanks.

**Comment by mfawzysami** [comment]
@kgururaj  I strongly recommend you to look at "Querying Data" section in GenomicsDB Wiki, I am lost there, It has been written by a person who was explaining to himself or as if he wanted to remember something by himself who knows GenomicsDB by heart, but for newcomers like us, it is a mystery.

In addition to this, Importing Data section also, the relationship between vid-mapping, call sets, and config JSON files is unclear, I struggled for days to format them clearly.

**Comment by mfawzysami** [comment]
@kgururaj @mlathara I am receiving this exception for a large file , what should i do , how can i increase the buffer size,

what():  File2TileDBBinaryException : read_one_line_fully && "Buffer did not have space to hold a line fully - increase buffer size"7216
